### PR TITLE
Optionally split the CFG pair into two right-sized forwards

### PR DESCRIPTION
**Pull request by badrabbitt**

### CODE REVIEW

**Comment by bourgois** [comment]
Some independent data for this PR, from profiling 0.2.0 for a production TTS deployment on an RTX A4500. We arrived at the same change from the other direction — measuring where time-to-first-audio goes — and this is the largest single item we found.

**The width is real compute, not launch overhead.** LLM forward per denoising step, CUDA events, no host sync, `num_step=16`, one voice-cloning request:

| jobs `B` | forward batch dim `2B` | forward / step |
|---|---|---|
| 1 | 2 | 31.18 ms |
| 2 | 4 | 57.76 ms |
| 4 | 8 | 116.04 ms |
| 8 | 16 | 228.44 ms |

Doubling the batch dim doubles the forward (1.85× / 2.01× / 1.97×). Fitting the top two points gives **14.05 ms per unit of batch width plus 3.6 ms fixed**, which predicts the `2B = 2` point at 31.7 against 31.18 measured. So the unconditional half is paying full price, and on this card the forward is **compute-bound rather than launch-bound**.

**Which is why the CUDA default here is worth reconsidering.** `split_cfg = self.device.type != "cuda"` turns this off exactly where our measurement says the trade is favourable. The PR's reasoning — tokens versus kernel launches — is right, but "CUDA ⇒ launch-bound" does not hold at these shapes on this class of card. A default keyed on the measured ratio (e.g. enable when `max_u_len / max_c_len` is below some threshold) would be better than one keyed on device type, and at minimum the CUDA default deserves a number behind it.

**How much the padding is, in voice-cloning mode.** For our request (German, 16 steps, a 7.14 s reference, ~5.9 s of target audio) `u_len ≈ 148` against `c_len ≈ 365–443`, i.e. the unconditional rows carry **~1/3** of the columns they are padded to. Model dims are 28 layers, hidden 1024, 16 heads × head_dim 128, intermediate 3072, so the per-token cost is dominated by the linear term and a right-sized unconditional forward is ≈ 0.33× the conditional one. Through the fit above, this PR is worth **roughly 90–148 ms of a 616 ms TTFA (15–24 %)** for us — the band is whether the 3.6 ms intercept is per-launch (paid twice here) or per-step (paid once), which we have not yet separated. This is an estimate from the fit, not a measurement of your patch; I will follow up with the direct timing.

**On correctness, your comment states it exactly right and I would put it more prominently in the PR body**, because it is the claim a reviewer has to check: the padded positions only ever attend to themselves, and there is no batch-coupled operation anywhere in the forward (`_prepare_embed_inputs` is an embedding lookup plus elementwise ops, and the LLM is a stock Qwen3), so the surviving rows' logits cannot change. The result is unchanged **by construction**. The only residual is GPU kernel-selection nondeterminism at a different batch shape — a numerical-tolerance question, not a sampling one. That distinction is what separates this PR from "turn CFG off", which genuinely does change what the model generates.

**One review note, on the non-split path.** The `else` branch now does three `torch.cat` calls *per denoising step* to rebuild the `2B` tensors that the current code allocates once and mutates in place:

```python
input_ids=torch.cat([cond_input_ids, uncond_input_ids], dim=0),
audio_mask=torch.cat([cond_audio_mask, uncond_audio_mask], dim=0),
attention_mask=torch.cat([cond_attention_mask, uncond_attention_mask], dim=0),
```

It is small at `B = 1` (the attention mask is a `(2,1,S,S)` bool, ~350 KB at `S ≈ 419`) but it is `O(B · max_c_len²)` and it lands on the **default** path, so every existing user pays it to keep current behaviour. Hoisting the concatenated tensors out of the step loop and writing the per-step token updates into them — as the current code already does — would make the non-split path allocation-free again and remove the objection.

Happy to contribute the direct measurement if that would help move this along.

**Comment by bourgois** [comment]
**Retracting the main claim in my comment above: I was wrong, and your CUDA default is right.**

I argued that `split_cfg = self.device.type != "cuda"` disables this patch "exactly where our measurement says the trade is favourable", on the grounds that our forward is compute-bound. I have now measured the patch's actual shapes instead of extrapolating a fit, and it goes the other way.

**The measurement.** Same card (RTX A4500), omnivoice 0.2.0, fp16, sdpa. Tensors built the way `_generate_iterative` builds them, `forward` called directly, CUDA events, no `generate()`. Our voice-cloning request is `c_len = 394`, `u_len = 145`.

Forward time by batch dim, at `c_len = 394`:

| batch dim | 1 | 2 | 3 | 4 | 6 | 8 |
|---|---|---|---|---|---|---|
| ms | **27.9** | **29.2** | 48.4 | 58.5 | 87.6 | 115.7 |

It is linear at ~13.5 ms per unit **from 3 upward** — that is the region my earlier numbers came from, and the fit there is fine. **Below 3 it saturates.** Widths 1 and 2 cost the same. At the unconditional length the plateau runs further still: at `len = 145`, widths 1/2/3/4 cost 26.6 / 27.7 / 28.1 / 27.9 ms, and only width 8 moves (42.1). So on this card the forward is **occupancy-bound at small width × length**, and the unconditional row of a `B = 1` CFG pair is very nearly free.

**What that does to this patch at `B = 1`.** Paired per rep over 101 reps against the status quo `(2, 394)`, since the box is shared and two identical arms drift ~2 ms:

| arm | Δ per step | over 16 steps |
|---|---|---|
| the same `(2,394)` arm again (instrument floor) | −0.51 ms | — |
| `(1, 394)` — drop the unconditional half entirely | −0.75 ms | ≈ 4 ms |
| **`(1, 394)` then `(1, 145)` — this patch** | **+28.27 ms** | **+452 ms** |

Each forward pays the ~27 ms floor in full, so splitting one launch into two roughly doubles the step. For a single-request latency workload on this card the patch is a large regression, and gating it off on CUDA is the correct default. My apologies for the noise.

**Two things I would still offer, because I do not think this makes the patch wrong.**

1. **The padding it removes is real and large.** Measured across six of our voice-cloning cells, `u_len / c_len` runs **0.257 to 0.371** — the unconditional rows are padded to between 2.7× and 3.9× their own length. On hardware or at batch sizes where the forward is genuinely compute-bound, that is exactly the saving you describe. Our own numbers say the linear region starts at batch dim 3, so **a server batching two or more requests is already in it**, and that is where this patch should pay. I only measured `B = 1`.

2. **Which suggests the default could key on the shape rather than on the device.** Something like "split when `B * max_u_len` is above the occupancy knee" would enable it for the batched server case on CUDA while keeping `B = 1` on the fast path. That needs a threshold somebody measures per architecture, so device type is a reasonable proxy to ship — but the current default also silently disables the patch for the batched CUDA case it is best suited to, which may be worth a note in the docstring.

The `torch.cat` point in my earlier comment stands on its own: the non-split branch now rebuilds the `2B` tensors every denoising step rather than allocating once and mutating in place, which is `O(B · max_c_len²)` on the **default** path. Hoisting them out of the step loop would remove that cost for everyone who does not opt in.

Raw data and the probe are public in our repo if useful: `scripts/probe_omnivoice_cfg_width.py` and `docs/bench/11f3a2f-omnivoice-cfg-width.json` in Voxist/voxist-tts.
